presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java

```diff
@@ -13,40 +13,142 @@
+        }
+        Multimap<DynamicFilterId, Symbol> symbolsMap = builder.build();
+
+        // Iterate over dynamic filters that are collected (correspond to one of the futures), and required for filtering (correspond to one of the descriptors).
+        // It is possible that some dynamic filters are collected in a different stage - and will not available here.
+        // It is also possible that not all local dynamic filters are needed for this specific table scan.
+        Set<DynamicFilterId> filterIds = Sets.intersection(symbolsMap.keySet(), futures.keySet());
+        List<ListenableFuture<TupleDomain<ColumnHandle>>> predicateFutures = filterIds.stream()
+                .map(filterId -> {
+                    // Probe-side columns that can be filtered with this dynamic filter resulting domain.
+                    List<ColumnHandle> probeColumns = symbolsMap.get(filterId).stream()
+                            .map(probeSymbol -> requireNonNull(columnsMap.get(probeSymbol), () -> format("Missing probe column for %s", probeSymbol)))
+                            .collect(toImmutableList());
+                    return Futures.transform(
+                            futures.get(filterId),
+                            // Construct a probe-side predicate by duplicating the resulting domain over the corresponding columns.
+                            (domain) -> TupleDomain.withColumnDomains(
+                                    probeColumns.stream()
+                                            .collect(toImmutableMap(
+                                                    (column) -> column,
+                                                    (column) -> domain))),
+                            directExecutor());
+                })
+                .collect(toImmutableList());
+        return new TableSpecificDynamicFilter(predicateFutures);
+    }
+
+    // Table-specific dynamic filter (collects all domains for a specific table scan)
+    private static class TableSpecificDynamicFilter
+            implements DynamicFilter
+    {
+        @GuardedBy("this")
+        private CompletableFuture<?> isBlocked;
+
+        @GuardedBy("this")
+        private TupleDomain<ColumnHandle> currentPredicate;
+
+        @GuardedBy("this")
+        private int futuresLeft;
+
+        public TableSpecificDynamicFilter(List<ListenableFuture<TupleDomain<ColumnHandle>>> predicateFutures)
+        {
+            this.futuresLeft = predicateFutures.size();
+            this.isBlocked = isComplete() ? NOT_BLOCKED : new CompletableFuture();
+            this.currentPredicate = TupleDomain.all();
+            predicateFutures.stream().forEach(future -> addSuccessCallback(future, this::update, directExecutor()));
+        }
+
+        private synchronized void update(TupleDomain<ColumnHandle> predicate)
+        {
+            this.futuresLeft -= 1;
+            verify(this.futuresLeft >= 0);
+            verify(this.isBlocked.complete(null));
```

Comment:
this should be executed outside of `synchronization` section, see: `io.prestosql.operator.exchange.LocalExchangeSource#addPage`

```
        ...
        // notify readers outside of lock since this may result in a callback
        notEmptyFuture.set(null);
```

it's a good practice to do callbacks outside of synchronization section